### PR TITLE
P2374R4: `views::cartesian_product`

### LINKED ISSUE #2923: P2374R4 `views::cartesian_product`
[P2374R4](https://wg21.link/P2374R4) `views::cartesian_product`
[P2540R1](https://wg21.link/P2540R1) Empty Product For Certain Views
LWG-3760 `cartesian_product_view::`*`iterator`*'s *`parent_`* is never valid
LWG-3761 `cartesian_product_view::`*`iterator`*`::operator-` should pass by reference
LWG-3801 `cartesian_product_view::`*`iterator`*`::`*`distance-from`* ignores the size of last underlying range
LWG-3820 `cartesian_product_view::`*`iterator`*`::`*`prev`*  is not quite right
LWG-3849 `cartesian_product_view::`*`iterator`*'s default constructor is overconstrained

Feature-test macro (expected):
```cpp
#define __cpp_lib_ranges_cartesian_product 202207L
```

### PR DESCRIPTION
**Pull request by JMazurkiewicz**
## The final boss of C++23's `std::views`

Implemented papers:
* WG21-P2374R4,
* WG21-P2540R1,
* LWG-3760,
* LWG-3761,
* LWG-3801,
* LWG-3820,
* LWG-3849.

Closes #2923.

Drive-by: replace `consteval` with `_CONSTEVAL` in `views::adjacent_transform` implementation.

<details>
<summary>Implementation details (please read while reviewing)</summary>

For simplicity: `CPV` means `cartesian_product_view` here.

* Implementation of preconditions for `CPV::size` and `CPV::iterator::distance-from` functions:

  Preconditions of both of those functions require us to check if multiplication or addition is possible without overflow (signed or unsigned). Therefore this PR adds two functions for safe multiplication/addition:

  * `bool _Add_with_overflow_check(x, y, &out)` - adds two integers `x` and `y` without undefined behaviour, saves the result in `&out`, and if an overflow occured - returns true.
  * `bool _Multiply_with_overflow_check(x, y, &out)` - multiplies two integers `x` and `y` without undefined behaviour, saves the result in `&out`, and if an overflow occured - returns true. It is based on [`llvm::MulOverflow`](https://github.com/llvm/llvm-project/blob/cd935224c02fd65719ae9fd42dc89bf7b599697f/llvm/include/llvm/Support/MathExtras.h#L725-L750) function (I hope this is fine, since [`NOTICE.txt` mentions LLVM project](https://github.com/microsoft/STL/blob/daa994bfc41c36196c536f2b68388f859d6bd656/NOTICE.txt#L133-L139)).

  Signatures of both of those functions are based on Clang's intrinsics `__builtin_add_overflow` and `__builtin_mul_overflow`, except `out` parameter is a reference (not a pointer).

* Implementation of `iterator::operator+=`'s preconditions [[range.cartesian.iterator]/19](http://eel.is/c++draft/range.cartesian#iterator-19):

  Currently advacing past end is detected only when first range models `ranges::sized_range`. It is possible do detect it for any range `Rng`, but it would possibly make `operator+=` linear (for the last range):

  ```c++
  // How we can check advancement past end if Rng does not model `ranges::sized_range`:
  const auto _End = _RANGES end(_Range);

  // vvv THIS MIGHT BE LINEAR OPERATION vvv
  const auto _Dist = _RANGES advance(_It, static_cast<iter_difference_t<_Iter>>(_Off), _End);

  // If `_Dist < _Off` then it would have been "past the end" advancement (if we have used `operator+=`).
  _STL_VERIFY(_Dist == _Off, "{message}");

  // If we reached the end, make sure other iterators are at the begin (otherwise we are "past the end").
  if (_It == _End) {
    _STL_VERIFY(_Entire_tail_at_begin(make_index_sequence<sizeof...(_Rest)>{}), "{message}");
  }
  ```

* Implementation of recommended practice for `range_size_t` and `range_difference_t`:

  This PR adds `_Cartesian_product_optimal_size_type` function, which **tries** to find the best unsigned integer type that satisfies recommendation [[range.cartesian.view]/8](http://eel.is/c++draft/range.cartesian#view-8):

  > Recommended practice: The return type should be the smallest unsigned-integer-like type that is sufficiently wide to store the product of the maximum sizes of all the underlying ranges, if such a type exists.

  * To do so, I've added `_Compile_time_max_size` variable, which detects types with max size known at compile time. Currently those types are:

    * `[const] span<T, N>`, where `N` is not `dynamic_extent`,
    * `[const] array<T, N>`,
    * any range with static `size` function, like `views::empty` or `views::single` (`ranges::tiny-range` concept does similar thing),
    * `ranges::ref_view` and `ranges::owning_view`.

  * Then `_Cartesian_product_max_size_bit_width<Rng>` function uses this variable to find the smallest bit width for integer type that is able to hold max possible size of given range.

  * Then `_Compile_time_max_size<Ranges...>` function sums the results of invocation `_Cartesian_product_max_size_bit_width` for each range in `Ranges`. This sum is the minimal bit width of integer type that is able to hold sum or product of max possible sizes of all ranges in `Ranges...` pack.

  * Then we choose our minimal size type:

    ```c++
    if constexpr (_Optimal_size_type_bit_width <= 8) {
        return uint_least8_t{};
    } else if constexpr (_Optimal_size_type_bit_width <= 16) {
        return uint_least16_t{};
    } else if constexpr (_Optimal_size_type_bit_width <= 32) {
        return uint_least32_t{};
    } else if constexpr (_Optimal_size_type_bit_width <= 64) {
        return uint_least64_t{};
    } else {
        return _Unsigned128{}; // This might not be able to hold product/sum of max possible sizes
    }
    ```

  * The `range_size_t` of our `CPV` is common type of chosen minimal size type and `range_size_t`s of underlying ranges.

  That said, let's look at the benefits of such approach:

  ```c++
  // We known that max size of `arr` is always 4. Therefore we know `size_t` is able to hold result of `cartesian_product_view::size`:
  array arr{1, 2, 3, 4};
  auto v1 = views::cartesian_product(arr, arr, arr);
  // ^^^ range_size_t == size_t, range_difference_t == ptrdiff_t ^^^

  // We don't know what is the max size of `vec` at compile time. Therefore we have to choose bigger size type for our cartesian product:
  vector vec{1, 2, 3, 4};
  auto v2 = views::cartesian_product(vec, vec, vec);
  // ^^^ range_size_t == _Unsigned128, range_difference_t == _Signed128 ^^^
  ```

  The usage of `_Compile_time_max_size` may be (and should be) expanded (if we decide to use this approach). There are much more ranges with max size known at compile time, like `vector<T, allocator<T>>` (value returned by `allocator<T>::max_size` is known at compile time).
</details>

### CODE REVIEW

File: stl/inc/ranges [hunk 1]
```diff
@@ -9169,6 +9169,632 @@ namespace ranges {
...
+                return _Sum;
+#else // ^^^ _ITERATOR_DEBUG_LEVEL != 0 / _ITERATOR_DEBUG_LEVEL == 0 vvv
+                return (_Scaled_distance<_Indices>(_Tpl) + ...);
+#endif //_ITERATOR_DEBUG_LEVEL == 0
+            }
+
+            template <class _Tuple>
+            _NODISCARD constexpr _Difference_type<_Const> _Distance_from(const _Tuple& _Tpl) const {
+                return _Scaled_sum(_Tpl, make_index_sequence<1 + sizeof...(_Rest)>{});
+            }
```
Comment (tylerbrawl) [inline stl/inc/ranges:9494]: I was working on this at some point, and I think we can reduce the entire implementation of `_Distance_from()` to the following **(sans overflow checks)**:

```C++
template <size_t _Index, class _Tuple>
_NODISCARD constexpr difference_type _Distance_from_impl(const _Tuple& _Rhs_tuple) const noexcept {
	const difference_type _Curr_diff =
		static_cast<difference_type>(_STD get<_Index>(_Current) - _STD get<_Index>(_Rhs_tuple));

        if constexpr (_Index > 0) {
		const difference_type _Curr_size =
                        static_cast<difference_type>(_RANGES size(_STD get<_Index>(_Parent->_Base)));
                return _Curr_diff + _Curr_size * _Distance_from_impl<_Index - 1>(_Rhs_tuple);
        } else {
                return _Curr_diff;
        }


template <class _Tuple>
_NODISCARD constexpr difference_type _Distance_from(const _Tuple& _Rhs_tuple) const noexcept {
	// This recursive implementation is a mathematically simplified version of the distance-from
	// function as specified in N4928: [range.cartesian.iterator]/7. It can drastically reduce the
	// number of multiplication operations when the number of views is high.

	return _Distance_from_impl<sizeof...(_Rest)>(_Rhs_tuple);
}
```

This works by simplifying the expression represented by the `_Distance_from()` function. For instance, consider the case where `sizeof...(_Rest) == 2`. Let `diff_N` represent `static_cast<difference_type>(std::get<N>(_Current) - std::get<N>(_Rhs_tuple))` and `size_N` represent `static_cast<difference_type>(ranges::size(std::get<N>(_Parent->_Bases)))`. The returned `scaled-sum` value is calculated as follows:

```
scaled-sum = scaled-distance(0) + scaled-distance(1) + scaled-distance(2)
           = (diff_0 * scaled-size(1)) + (diff_1 * scaled-size(2)) + (diff_2 * scaled-size(3))
           = (diff_0 * size_1 * size_2 * 1) + (diff_1 * size_2 * 1) + (diff_2 * 1)
           = diff_2 + size_2 * (diff_1 + size_1 * (diff_0))
```

As can be seen, this drastically reduces the number of required multiplication operations and should thus be much faster. It's easy to then generalize this to arbitrary values for the `sizeof...(_Rest)`, which is precisely what `_Distance_from_impl()` does.

File: stl/inc/ranges [hunk 2]
```diff
@@ -9169,6 +9169,632 @@ namespace ranges {
...
+
+    template <class _Rng>
+    concept _Cartesian_product_common_arg = common_range<_Rng> || (sized_range<_Rng> && random_access_range<_Rng>);
+
+    template <bool _Const, class _First, class... _Rest>
+    concept _Cartesian_product_is_bidirectional = (bidirectional_range<_Maybe_const<_Const, _First>> && ...
+                                                   && (bidirectional_range<_Maybe_const<_Const, _Rest>>
+                                                       && _Cartesian_product_common_arg<_Maybe_const<_Const, _Rest>>) );
+
+    template <class _First, class... _Rest>
```
Comment (CaseyCarter) [inline stl/inc/ranges:9238]: Quick drive-by since I just suggested removing the unused template parameter pack from the Standard: remove `class... _Rest` here ;) There are only two uses to cleanup, and there's a minimal throughput benefit.

**Comment by StephanTLavavej** [comment]
Is this ready for initial review?

File: stl/inc/ranges (hunk 1 above)
Comment (JMazurkiewicz) [reply stl/inc/ranges:9494]: Thanks! I've added your *`distance-from`* implementation with extra overflow checks!

**Comment by JMazurkiewicz** [comment]
> Is this ready for initial review?

Yes, it is.

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `main` to resolve adjacent-add conflicts with #3546. This adds both feature-test macros in sorted order (with clang-format column alignment) and places `adjacent_transform` before `cartesian_product` to match the Standard's order.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! I reviewed all of the product code; I'll review the test code later.

File: stl/inc/ranges [hunk 3]
```diff
@@ -9553,6 +9553,595 @@ namespace ranges {
...
+
+    template <class _Ty>
+    inline constexpr auto _Compile_time_max_size<owning_view<_Ty>> = _Compile_time_max_size<_Ty>;
+
+    template <class _Ty>
+        requires sized_range<const _Ty>
+    inline constexpr auto _Compile_time_max_size<const owning_view<_Ty>> = _Compile_time_max_size<const _Ty>;
+
+    template <range _Rng>
+    _NODISCARD consteval int _Cartesian_product_max_size_bit_width() noexcept {
```
[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply stl/inc/ranges:9683]: > Are these "nice to have" cases, where the `consteval` keyword isn't absolutely required?

I think there is only one now in `<random>` which is needed to be available in C++14 (with TR1).
https://github.com/microsoft/STL/blob/5116678ea11245b232a6e43824aad25fb60b9c6e/stl/inc/random#L1002

All other uses of `_CONSTEVAL` are only in C++20/23 modes, and there's already unprotected uses of `consteval` in `<ranges>`.
https://github.com/microsoft/STL/blob/5116678ea11245b232a6e43824aad25fb60b9c6e/stl/inc/ranges#L9233

File: stl/inc/ranges [hunk 4]
```diff
@@ -9553,6 +9553,595 @@ namespace ranges {
...
+                auto& _Range = _STD get<_Index>(_Parent->_Bases);
+                auto& _It    = _STD get<_Index>(_Current);
+                using _Rng   = remove_reference_t<decltype(_Range)>;
+                using _Iter  = remove_reference_t<decltype(_It)>;
+
+                if constexpr (_Index > 0) {
+                    _STL_INTERNAL_STATIC_ASSERT(sized_range<_Rng>);
+                    const auto _Size       = static_cast<_Difference_type<_Const>>(_RANGES ssize(_Range));
+                    const auto _Begin      = _RANGES begin(_Range);
+                    const auto _It_off     = static_cast<_Difference_type<_Const>>(_It - _Begin);
```
[the remark this answers is not in the document]
Comment (JMazurkiewicz) [reply stl/inc/ranges:9786]: `_Range` is requires to be random-access here, since this is implementation of `operator+=`:
```c++
template <bool _Const, class _First, class... _Rest>
concept _Cartesian_product_is_random_access = (random_access_range<_Maybe_const<_Const, _First>> && ...
                                               && (random_access_range<_Maybe_const<_Const, _Rest>>
                                                   && sized_range<_Maybe_const<_Const, _Rest>>) );
```
I'm just going to remove this internal assertion, it is unnecessary.

File: stl/inc/ranges (hunk 3 above)
[the remark this answers is not in the document]
Comment (JMazurkiewicz) [reply stl/inc/ranges:9683]: I've replaced all 4 occurrences of `consteval` with `_CONSTEVAL`.

File: stl/inc/ranges [hunk 5]
```diff
@@ -9553,6 +9553,595 @@ namespace ranges {
...
+    _NODISCARD constexpr bool _Multiply_with_overflow_check(_Int _Left, _Int _Right, _Int& _Out) {
+#ifdef __clang__
+        if constexpr (integral<_Int>) {
+            return __builtin_mul_overflow(_Left, _Right, &_Out);
+        } else
+#endif // __clang__
+        {
+            if constexpr (!_Signed_integer_like<_Int>) {
+                _Out = _Left * _Right;
+                return _Left != 0 && _Right != 0 && (_Out < _Left || _Out < _Right);
```
[the remark this answers is not in the document]
Comment (JMazurkiewicz) [reply stl/inc/ranges]: This is incorrect, I've change overflow detection method to this:
```c++
return _Left != 0 && _Right > (numeric_limits<_Int>::max)() / _Left;
```

**Comment by StephanTLavavej** [comment]
Awesome, thanks - I'll resume reviewing 😻

File: tests/std/tests/P2374R4_views_cartesian_product/test.cpp [hunk 6]
```diff
@@ -0,0 +1,1030 @@
...
+    template <class R>
+    static constexpr void call() {
+        typename R::template type<const int> r0{get<0>(some_ranges)};
+        test_one(get<0>(expected_results), r0);
+
+        if constexpr (ranges::forward_range<typename R::template type<const int>>) {
+            typename R::template type<const int> r1{get<1>(some_ranges)};
+            test_one(get<1>(expected_results), r0, r1);
+
+#if !(defined(__clang__) && defined(_DEBUG) && defined(_MT) && !defined(_DLL)) // constexpr limit
```
Comment (StephanTLavavej) [inline tests/std/tests/P2374R4_views_cartesian_product/test.cpp:876]: `defined(_MT)` is a red herring here (it's always defined; a relic of when we had single-threaded and multi-threaded CRTs).

File: tests/std/tests/P2374R4_views_cartesian_product/test.cpp [hunk 7]
```diff
@@ -0,0 +1,1030 @@
...
+                  <= sizeof(size_t));
+    STATIC_ASSERT(
+        sizeof(range_difference_t<cartesian_product_view<ref_view<Arr>, ref_view<const Arr>, owning_view<Arr>>>)
+        <= sizeof(ptrdiff_t));
+
+    // One vector should not use big integer-class type...
+    STATIC_ASSERT(sizeof(range_size_t<cartesian_product_view<all_t<Vec>>>) <= sizeof(size_t));
+    STATIC_ASSERT(sizeof(range_difference_t<cartesian_product_view<all_t<Vec>>>) <= sizeof(ptrdiff_t));
+
+    // ...but two vectors will
```
Comment (StephanTLavavej) [inline tests/std/tests/P2374R4_views_cartesian_product/test.cpp:977]: Nitpick: Would prefer a space after the ellipsis here.

File: tests/std/tests/P2374R4_views_cartesian_product_death/test.cpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+void test_view_size() {
+    auto v = views::cartesian_product(much_ints, much_ints, much_ints);
+    // Size of cartesian product cannot be represented by _Size_type<false> (N4928 [range.cartesian.view]/10).
+    (void) v.size();
+}
+
+void test_view_const_size() {
+    auto v = views::cartesian_product(much_ints, much_ints, much_ints);
+    // Size of cartesian product cannot be represented by _Size_type<true> (N4928 [range.cartesian.view]/10).
+    (void) as_const(v).size();
```
Comment (StephanTLavavej) [inline tests/std/tests/P2374R4_views_cartesian_product_death/test.cpp:27]: Technically should include `<utility>` for `as_const()`.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Finished reviewing, pushed changes for all of the issues I found - the most significant being bogus code in `is_iter_swap_nothrow` that was missed because nothing was exercising the codepath. With this, I think the final boss is ready for final review - thanks again! :heart_eyes_cat: :tada:

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `main` to resolve trivial adjacent-add conflicts in `yvals_core.h`.

File: stl/inc/ranges [hunk 9]
```diff
@@ -9553,6 +9553,604 @@ namespace ranges {
...
+
+            template <size_t... _Indices>
+            _NODISCARD constexpr bool _Is_end(index_sequence<_Indices...>) const {
+                return ((_STD get<_Indices>(_Current) == _RANGES end(_STD get<_Indices>(_Parent->_Bases))) || ...);
+            }
+
+            template <size_t... _Indices>
+            _NODISCARD constexpr auto _End_tuple(index_sequence<_Indices...>) const {
+                return tuple{_RANGES end(_STD get<0>(_Parent->_Bases)),
+                    _RANGES begin(_STD get<_Indices + 1>(_Parent->_Bases))...};
```
Comment (cpplearner) [inline stl/inc/ranges:9829]: This CTAD might break in a pathological case, namely when the result of `end` is a `std::pair` that provides everything required by `input_iterator`.

IMO it's safer to write out the template arguments in generic code.

Occurs below in `cartesian_product_view::end`.

File: stl/inc/ranges [hunk 10]
```diff
@@ -9553,6 +9553,595 @@ namespace ranges {
...
+
+    template <_Integer_like _Int>
+    _NODISCARD constexpr bool _Multiply_with_overflow_check(_Int _Left, _Int _Right, _Int& _Out) {
+#ifdef __clang__
+        if constexpr (integral<_Int>) {
+            return __builtin_mul_overflow(_Left, _Right, &_Out);
+        } else
+#endif // __clang__
+        {
+            if constexpr (!_Signed_integer_like<_Int>) {
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/ranges:9583]: we could also say _Is_unsigned_integer_like or soemting, but tbh I strongly prefer just keeping this identical to what's in LLVM.

**Review by barcharcraz — CHANGES_REQUESTED** [review]

File: tests/std/tests/P2374R4_views_cartesian_product/test.cpp [hunk 11]
```diff
@@ -0,0 +1,1032 @@
...
+using namespace std;
+using ranges::bidirectional_range, ranges::random_access_range, ranges::sized_range, ranges::common_range,
+    ranges::iterator_t;
+
+template <bool B, class T>
+using maybe_const = conditional_t<B, const T, T>;
+
+// Helper concepts from [range.cartesian.view]
+template <bool IsConst, class First, class... Rest>
+concept CartesianProductIsRandomAccess = (random_access_range<maybe_const<IsConst, First>> && ...
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply tests/std/tests/P2374R4_views_cartesian_product/test.cpp:28]: I believe the rationale is to double-check the implementation against an independent translation of the Standard's exposition-only concepts into code, so that if the implementation is modified in the future, we could notice any discrepancies. If we directly use `_Ugly` internals in test code (which we try to avoid doing, although not 100% strictly), we lose any chance of noticing such behavioral differences.

File: stl/inc/ranges [hunk 12]
```diff
@@ -9553,6 +9553,604 @@ namespace ranges {
...
+        }
+    }
+
+    template <_Integer_like _Int>
+    _NODISCARD constexpr bool _Multiply_with_overflow_check(const _Int _Left, const _Int _Right, _Int& _Out) {
+#ifdef __clang__
+        if constexpr (integral<_Int>) {
+            return __builtin_mul_overflow(_Left, _Right, &_Out);
+        } else
+#endif // __clang__
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/ranges:9581]: I've filed DevCom-10326281 "Consider implementing Clang's checked arithmetic builtins".

File: tests/std/tests/P2374R4_views_cartesian_product/test.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply tests/std/tests/P2374R4_views_cartesian_product/test.cpp:28]: Yeah, I don't have a huge problem with this, but it's not a super high value test, esp since the concepts are ugly so their behavior doesn't matter as long as everything else works.

**Review by barcharcraz — APPROVED** [review]

File: tests/std/tests/P2374R4_views_cartesian_product/test.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply tests/std/tests/P2374R4_views_cartesian_product/test.cpp:28]: Thanks. These pretty concepts are just used to check the observable properties that are controlled by the ugly concepts. I agree that the test coverage value is relatively low (these fundamentals are unlikely to break), but it's consistent with our usual view testing where we try to inspect all of the major observable properties, and even low value coverage ends up being valuable in large quantities (as we have seen where later compiler or library features end up finding surprise interactions in simple pieces of test code that had always previously passed).

**Comment by StephanTLavavej** [comment]
I've pushed a conflict-free merge with `main`, followed by a workaround for the notorious #1030 and a perma-workaround for the newly reported VSO-1782194 "x86chk `/analyze` emits bogus message '`AST_FUNCDEFN`: function should be compiler-generated: `__vec_copy_ctor`'". The static analyzer really disliked the `tuple` of `array`s of `tuple`s, and it didn't serve a fundamental purpose, so I've split it up into individual `array`s (and I've also simplified their initialization by providing explicit sizes).

**Comment by StephanTLavavej** [comment]
Final boss defeated! :tada: :heart_eyes_cat: :rocket:
